File: android/app/src/main/java/com/quartier/quartier/profile/ProfileViewModel.kt
package com.quartier.quartier.profile

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.quartier.quartier.ImageRepository
import com.quartier.quartier.R
import com.quartier.quartier.database.Socials
import com.quartier.quartier.database.SocialsRepository
import com.quartier.quartier.database.SupabaseException
import com.quartier.quartier.database.UploadStatus
import com.quartier.quartier.database.User
import com.quartier.quartier.database.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.exceptions.HttpRequestException
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

//View model for the profile screen, used to fetch the user's profile

data class ProfileScreenState(
    val user: User?,
    val socials: Socials?,
    val isRefreshing: Boolean = false,
    val userMessage: Int?
)

data class EditProfileState(
    val shown: Boolean = false,
    val newPfpUri: Uri?,
    val saveStatus: UploadStatus?
)

@HiltViewModel
class ProfileViewModel @Inject constructor(
    private val userRepository: UserRepository,
    private val socialsRepository: SocialsRepository,
    private val imageRepository: ImageRepository
) : ViewModel() {
    private val _user: MutableStateFlow<User?> = MutableStateFlow(null)
    private val _socials: MutableStateFlow<Socials?> = MutableStateFlow(null)
    private val _isRefreshing: MutableStateFlow<Boolean> = MutableStateFlow(false)

    private val _userMessage = MutableStateFlow<Int?>(null)

    val uiState: StateFlow<ProfileScreenState> = combine(_user, _socials, _isRefreshing, _userMessage) { user, socials, isRefreshing, message ->
        ProfileScreenState(user, socials, isRefreshing, message)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000),
        ProfileScreenState(null, null, false, null))

    private val _editShown: MutableStateFlow<Boolean> = MutableStateFlow(false)
    private val _newPfpUri: MutableStateFlow<Uri?> = MutableStateFlow(null)
    private val _saveStatus: MutableStateFlow<UploadStatus?> = MutableStateFlow(null)

    val editState: StateFlow<EditProfileState> = combine(_editShown, _newPfpUri, _saveStatus) { show, uri, status ->
        EditProfileState(show, uri, status)
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000),
        EditProfileState(false, null, null))

    init {
        refreshUser()
    }

    fun refreshUser() {
        viewModelScope.launch {
            _isRefreshing.value = true
            try {
                _user.value = userRepository.getUser()
            } catch (e: SupabaseException) {
                _userMessage.value = R.string.no_internet
            } catch (e: HttpRequestException) {
                _userMessage.value = R.string.no_internet
            }
            _isRefreshing.value = false
        }
        viewModelScope.launch {
            try {
                _socials.value = socialsRepository.getUserSocials()
            } catch (e: SupabaseException) {
                _userMessage.value = R.string.no_internet
            } catch (e: HttpRequestException) {
                _userMessage.value = R.string.no_internet
            }
        }
    }

    @OptIn(ExperimentalUuidApi::class)
    fun saveUser(newName: String, newJob: String, newLinkedin: String) {
        //If the user didn't change any of these values, keep the current one
        _user.value?.let { user ->
            val savedName = newName.ifBlank { user.name }
            val savedJob = newJob.ifBlank { user.job }
            val savedLinkedin = newLinkedin.ifBlank { _socials.value?.linkedin_url }
            _saveStatus.value = UploadStatus.Loading

            viewModelScope.launch {
                val success = try {
                    coroutineScope {
                        //Save on all tables simultaneously
                        val profileJob = async { userRepository.updateUser(savedName, savedJob) }
                        val pfpJob = async {
                            _newPfpUri.value?.let { uri ->
                                val croppedImage = imageRepository.cropImageTo400(uri = uri) //Crop the pfp to the right format
                                val imageData = imageRepository.convertToWebPByteArray(croppedImage) //Convert to WebP
                                userRepository.uploadPfp(Uuid.random().toString() + ".webp", imageData) //Upload the image
                            }
                        }
                        val socialsJob = async {
                            savedLinkedin?.let { savedLinkedin ->
                                socialsRepository.upsertSocials(savedLinkedin)
                            }
                        }

                        awaitAll(profileJob, pfpJob, socialsJob)
                    }
                    true
                } catch (e: Exception) {
                    _saveStatus.value = UploadStatus.Error
                    false
                } catch (e: HttpRequestException) {
                    _saveStatus.value = UploadStatus.Error
                    false
                }

                if(success) _saveStatus.value = UploadStatus.Success
            }
        }
    }

    //Method to check if the link inserted is a valid linkedin profile
    fun matchesLinkedinRegex(input: String): Boolean {
        val linkedinRegex = Regex("^https://www\\.linkedin\\.com/in/[^/]+/?$")
        return linkedinRegex.matches(input)
    }

    fun previewNewPfp(uri: Uri) {
        _newPfpUri.value = uri
    }

    fun snackbarMessageShown() {
        _saveStatus.value = null
    }

    fun showEdit() {
        _editShown.value = true
    }

    fun hideEdit() {
        _editShown.value = false
    }
}

File: ios/Quartier/Profile/ProfileViewModel.swift
//
//  ProfileViewModel.swift
//  Connect
//
//  Created by Emile Turcotte on 2025-07-06.
//

import Foundation
import PhotosUI

//Viewmodel for the profile screen, fecthes the user's profile
@MainActor
protocol ProfileViewModel : ObservableObject {
    var userProfile: User? { get }
    var userSocials: Socials? { get }
    var userMessage: String? { get }
    var dismissEdit: Bool { get }
    
    func fetchProfile()
    func saveUser(name: String, jobTitle: String, linkedInURL: String, pfp: UIImage?)
    func matchesLinkedinRegex(input: String) -> Bool
}

class ProfileViewModelImpl: ProfileViewModel {
    private var userRepository: any UserRepository
    private var socialsRepository: any SocialsRepository
    private var imageRepository: any ImageRepository
    
    @Published var userProfile: User? = nil
    @Published var userSocials: Socials? = nil
    @Published var userMessage: String? = nil
    @Published var dismissEdit: Bool = false
    
    init(userRepository: any UserRepository = UserDatabase.shared, socialsRepository: any SocialsRepository = SocialsDatabase.shared, imageRepository: any ImageRepository = ImageManager()) {
        self.userRepository = userRepository
        self.socialsRepository = socialsRepository
        self.imageRepository = imageRepository
    }
    
    func fetchProfile() {
        Task {
            userProfile = try await userRepository.getUser()
        }
        Task {
            userSocials = try await socialsRepository.getUserSocials()
        }
    }
    
    func saveUser(name: String, jobTitle: String, linkedInURL: String, pfp: UIImage?) {
        guard let userProfile else { return }
        
        userMessage = "Saving..."
        
        //If the user didn't change it, keep it the way it was
        let savedName = name.isEmpty ? userProfile.name : name
        let savedJobTitle = jobTitle.isEmpty ? userProfile.job: jobTitle
        let savedLinkedIn = linkedInURL.isEmpty ? userSocials?.linkedin_url : linkedInURL
        
        Task {
            do {
                try await withThrowingTaskGroup(of: Void.self) { group in
                    //Save everything simultaneously
                    group.addTask {
                        try await self.userRepository.updateUser(name: savedName, jobTitle: savedJobTitle)
                    }
                    if let linkedin = savedLinkedIn {
                        group.addTask {
                            try await self.socialsRepository.upsertSocials(linkedInUrl: linkedin)
                        }
                    }
                    if let pfpImage = pfp {
                        group.addTask {
                            let scaledImage = await self.imageRepository.resizeTo400(image: pfpImage) //Crop image
                            let imageData = try await self.imageRepository.encodeToWebP(image: scaledImage) //Convert to WebP
                            try await self.userRepository.uploadPfp(fileName: UUID().uuidString+".webp", imageData: imageData) //Save image
                        }
                    }
                    
                    for try await _ in group {}
                    
                    return (Void())
                }
            }
            catch(let error) {
                print(error)
                userMessage = "An error happened saving your profile!"
                return
            }
            userMessage = "Profile saved successfully!"
            dismissEdit = true
            Task {
                try await Task.sleep(for: .seconds(5))
                userMessage = nil
                dismissEdit = false
            }
        }
    }
    
    nonisolated func matchesLinkedinRegex(input: String) -> Bool { //Check if the link is a valid linkedin link
        do {
            let regex = try Regex("^https://www\\.linkedin\\.com/in/[^/]+/?$")
            return try regex.wholeMatch(in: input) != nil
        } catch {
            return false
        }
    }
}
